# 错误

一个 [Sumor Cloud](https://sumor.cloud) 工具。  
[更多文档](https://sumor.cloud/error) 一个支持多语言、预定义错误代码和传递数据的错误处理库。

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Coverage](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Audit](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## 安装

```bash
npm i @sumor/error --save
```

## 先决条件

### Node.JS 版本

需要 Node.JS 版本 18.x 或更高版本

### 需要 Node.JS ES 模块

由于此软件包是用 ES 模块编写的，请在您的 `package.json` 文件中更改以下代码：

```json
{
  "type": "module"
}
```

## 用法

### 定义错误

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: '未找到用户',
    USER_EXISTED: '用户 {name} 已存在'
  }
})

throw new MyError('USER_NOT_FOUND')
// 输出: Error: 未找到用户

throw new MyError('USER_EXISTED', { name: 'Alice' })
// 输出: Error: 用户 Alice 已存在
```

### 多语言

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'en', // 默认语言
  code: {
    USER_NOT_FOUND: 'User not found',
    USER_EXISTED: 'User {name} existed'
  },
  i18n: {
    zh: {
      USER_NOT_FOUND: '用户未找到',
      USER_EXISTED: '用户 {name} 已存在'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'en' // 更改错误语言
console.log(error)
// 输出: Error: 用户 Alice 已存在

error.language = 'zh' // 更改错误语言
console.log(error)
// 输出: Error: 用户 Alice 已存在
```

### 将错误转换为 JSON

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: '未找到用户',
    USER_EXISTED: '用户 {name} 已存在'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// 输出: {"code":"USER_EXISTED","message":"用户 Alice 已存在"}
```

### 底层错误

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: '字段验证失败',
    FIELD_CANNOT_EMPTY: '字段 {name} 不能为空',
    FIELD_TOO_LONG: '字段 {name} 过长'
  },
  i18n: {
    zh: {
      FIELD_VERIFY_FAILED: '字段验证失败',
      FIELD_CANNOT_EMPTY: '字段 {name} 不能为空',
      FIELD_TOO_LONG: '字段 {name} 过长'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'username' }),
  new MyError('FIELD_TOO_LONG', { name: 'password' })
])

console.log(error.json())
/* 
输出: 
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"字段验证失败",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"字段 username 不能为空"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"字段 password 过长"
    }
  ]
}
*/

error.language = 'zh'
console.log(error.json())
/*
输出:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"字段验证失败",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"字段 username 不能为空"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"字段 password 过长"
    }
  ]
}
*/
```

### 结合标准错误

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: '字段验证失败',
    FIELD_CANNOT_EMPTY: '字段 {name} 不能为空',
    FIELD_TOO_LONG: '字段 {name} 过长'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'username' }),
  new MyError('FIELD_TOO_LONG', { name: 'password' }),
  new Error('未知错误')
])

console.log(error.json())
/*
输出:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"字段验证失败",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"字段 username 不能为空"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"字段 password 过长"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"未知错误"
    }
  ]
}
 */
```
